git-workflow.md

```diff
@@ -0,0 +1,77 @@
+# `git` Workflow
+
+## Terminology
+- `upstream` : This is the repo you want to contribute to. Ex: `github.com/konveyor/move2kube`
+- `origin` : This is your fork of the `upstream` repo. Ex: `github.com/myusername/move2kube`
+- `local` : This is a repo on your local machine (laptop/desktop). Usually this refers to the clone of your fork `origin`.
+
+## Setup
+This only needs to be done once:
+1. Fork the repo on Github.
+2. Clone your fork: `git clone <my fork url>`
+3. Add the `upstream` repo to the set of remote repos: `git remote add upstream <upstream url>`
+4. Check that you have both the `origin` and `upstream` set correctly: `git remote -v`
+
+## Debugging/Visibility
+When in doubt `git log --graph --all` and scroll around with the arrow keys.  
+This shows you the graph of commits and the commit each branch is pointing to.  
+If there are a lot of commits you can also add `--one-line` to make it easier to see.
+
+## Sync up
+Often our `local` repo and our fork `origin` will lag behind the main repo `upstream`.
+It is important to sync up with `upstream` before we rebase and submit a pull request on Github.
+
+1. Get the latest code from upstream: `git fetch upstream`
+2. Switch to the master branch: `git checkout master`
+3. Fast forward your local master branch to catch up with upstream: `git merge --ff-only upstream/master`
+4. Push the changes to your fork: `git push`
+
+## Making changes
+**ALWAYS BRANCH OUT OF MASTER**
+
+1. Switch to the master branch: `git checkout master`
+2. Create a new branch and check it out: `git checkout -b my-feature-or-bug-fix-branch`
+3. Make some changes.
+4. Add all changes to the staging area before committing: `git add -A`
+5. Commit and sign off on the commit: `git commit -s -m 'My commit message'`
+6. Push the new commits to your fork: `git push`
+7. Repeat steps 3 to 6 until you are ready to submit a pull request.
+
+## Submitting a pull request
+
+1. Make sure the code passes build and test: `make ci`
+2. Make sure all the changes are committed and the working tree is clean: `git status`
+3. Follow the steps to sync up with `upstream`.
+4. `git checkout my-feature-or-bug-fix-branch`
+5. Rebase onto the `upstream/master` branch: `git rebase upstream/master` . Fix any merge conflicts that occur.
+6. Make sure the rebased code passes build and test: `make ci`
+7. After a succcessful rebase push the changes to your fork: `git push --force`
+8. Submit a pull request on Github between your branch `my-feature-or-bug-fix-branch` on your fork `origin` and the `master` branch on `upstream`.
+
```

Comment:
Do you want to add a section on editing a pr with more commits or cleaning commits

Reply:
hmm, I REALLY don't want to explain rebasing.
rebasing requires some basic knowledge of `vi` or whatever text editor `git` chooses to use.
Also there are a LOT of different things you can do with `git rebase -i` and it can get confusing easily.